forbidigo/forbidigo.go

```diff
@@ -164,24 +192,44 @@ func (v *visitor) Visit(node ast.Node) ast.Visitor {
 		}
 		return v
 	case *ast.SelectorExpr:
+		selectorExpr = node
 	case *ast.Ident:
 	default:
 		return v
 	}
+
+	// The text as it appears in the source is always used because issue
+	// use that. The other texts to match against are extracted when needed
+	// by a pattern.
+	srcText := v.textFor(node)
+	pkgText := ""
 	for _, p := range v.linter.patterns {
-		if p.pattern.MatchString(v.textFor(node)) && !v.permit(node) {
-			v.issues = append(v.issues, UsedIssue{
-				identifier: v.textFor(node),
-				pattern:    p.pattern.String(),
-				pos:        node.Pos(),
-				position:   v.fset.Position(node.Pos()),
-				customMsg:  p.msg,
-			})
+		switch {
+		case p.matchPackage:
+			if pkgText == "" {
+				pkgText = v.textWithPackage(selectorExpr)
```

Comment:
I'd favor not having `textWithPackage` handle a nil selectorExpr because that pushes down some control flow making it harder to see what's happening.  Instead, we could determine whether `node` is a selectorExpr right here and if that is true and `p.matchPackage` is set, we could then swap in the alternate text.

Reply:
Ack.

Reply:
I've inlined both `match` and `textWithPackage`. As before, the code which resolves the package name silently ignores a pattern when any step along the way is not possible, now via `continue`. None of the current unit tests hit those conditions and I am still a bit unsure whether these should be treated as error instead.